### PR TITLE
Add more flexibility to BuckleComponent. Replaced ComponentMessage with ECS in BuckleComponent and ClimbComponent

### PR DESCRIPTION
**Pull request by Jackrost**
## About the PR 

*) Add more flexibility to BuckleComponent
*) ECS for BuckleComponent and ClimbComponent

:cl:
- add: Add more flexibility to BuckleComponent
- add: BuckleComponent and ClimbComponent using ECS now

### CODE REVIEW

**Comment by Jackrost** [comment]
Link it with issue #4265, please

File: Content.Shared/Buckle/SharedBuckleSystem.cs [hunk 1]
```diff
@@ -33,6 +37,18 @@ private void HandleDown(EntityUid uid, SharedBuckleComponent component, DownAtte
...
+        {
+            if (!component.Buckled) return;
+            args.Cancel();
+        }
+
+        private void HandleChangeDirection(EntityUid uid, SharedBuckleComponent component, ChangeDirectionAttemptEvent args)
+        {
+            if (!component.Buckled) return;
+            args.Cancel();
+        }
```
Comment (metalgearsloth) [inline Content.Shared/Buckle/SharedBuckleSystem.cs:50]: Would be better to probably have a separate "Buckled" component and then use the directed event for it, that way this stuff isn't being called for every single movement / rotation.

Downside is that RemoveComponent is currently bugged via networking so would need to check if it has a client version that runs too.

File: Content.Shared/Climbing/SharedClimbingSystem.cs [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+        {
+            base.Initialize();
+            SubscribeLocalEvent<SharedClimbingComponent, MovementAttemptEvent>(HandleMove);
+        }
+
+        private void HandleMove(EntityUid uid, SharedClimbingComponent component, MovementAttemptEvent args)
+        {
+            if (!component.GetIsTransitioning()) return;
+            args.Cancel();
+        }
```
Comment (metalgearsloth) [inline Content.Shared/Climbing/SharedClimbingSystem.cs:18]: Same as above. Also climbing checking IsTransitioning will be removed someday so not much point having this tbh.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

File: Content.Shared/Climbing/SharedClimbingSystem.cs (hunk 2 above)
Comment (metalgearsloth) [reply Content.Shared/Climbing/SharedClimbingSystem.cs:18]: Actually just slap a TODO on this and we'll deal with it later.

File: Content.Client/Buckle/BuckleComponent.cs [hunk 3]
```diff
@@ -13,7 +13,7 @@ public class BuckleComponent : SharedBuckleComponent
 
         public override bool Buckled => _buckled;
 
-        public override bool TryBuckle(IEntity? user, IEntity to)
+        public override bool TryBuckle(IEntity? user, IEntity to, bool check_range, float? range)
```
[the remark this answers is not in the document]
Comment (Jackrost) [reply Content.Client/Buckle/BuckleComponent.cs:16]: Like I have discussed about this earlier in discord, it was for optimisation. In Kitchen Spike I use range check (InRangeUnobstructed method) after success result of "do_after" to be sure. After that TryBuckle proccess happend and doing the same thing again. Also default buckle range is very small. For chair it's ok, but Kitchen Spike is solid and placed in corner, so you need pulling body  very close, and moving something with mouse click just too long procces for this.

[the remark this answers is not in the document]
Comment (metalgearsloth) [reply Content.Client/Buckle/BuckleComponent.cs:16]: You're better off always doing the check in buckle and not kitchenspike then, no?

**Comment by metalgearsloth** [comment]
Changelog should also be player-friendly. In this case I don't think they'd know what a refactor is so I'd probably delete it.

**Comment by metalgearsloth** [comment]
You coming back to this Jack?

**Comment by Jackrost** [comment]
> You coming back to this Jack?
Yeah,, but I still busy, so I complete it later

**Comment by metalgearsloth** [comment]
Been over 6 weeks so if you get time just get it re-opened.

**Pull request closed without merging**